source/agora/test/QuorumShuffle.d

```diff
@@ -133,7 +133,7 @@ unittest
             format("Node %s has quorum config %s. Expected: %s",
                 idx, node.getQuorumConfig(), quorums_4[idx])));
 
-    checkDistance(1);  // re-enrolled, starts from 1 again
+    checkDistance(0);  // re-enrolled, starts from 0 again
```

Comment:
Are you sure about that? The node needs to reveal the preimage of the commitment for the first block height. Unless there is something wrong with the implementation?

Reply:
I might be wrong though. It depends how `getPreimageAt` was implemented.

Reply:
So I guess for the initial first block the node validates, the preimage that is used is actually the commitment in the enrollment. Then it LGTM if that's the case.

Reply:
I just checked and the `validator_set` has the `preimage` distance set to zero on enrolling.

Reply:
That's normal. But I was wondering which preimage is taken for the first new block height in the call to `getRandomSeed` in the `EnrollmentManager`.

Reply:
Before creating a new block with a `proposed_height`, we need the pre-image which is at the distance of `proposed_height - enrolled_height - 1`. In this case, the required distance of pre-preimages should be 9 because the last pre-image is required at the end of the cycle. 
Please print the distance of the pre-images with the code as follows.
```
writeln("dist: ", nodes[0].getPreimage(enrollments[0].utxo_key).distance);
``` 
Additionally, new enrollments will be stored in the block of height 10.

Reply:
```
 auto results = this.db.execute(
                "SELECT preimage, enrolled_height, distance " ~
                "FROM validator_set WHERE key = ? AND enrolled_height <= ? " ~
                "AND enrolled_height + distance >= ?",
                enroll_key.toString(), height.value, height.value);
```
The first block is height 1 and the enrollment is at height 0. So distance must be at least 1.
At block 10 (assuming cycle 10) the enrollment is at height 10 and the height is also 10. So distance must be at least 0.
At block 11 the distance must be at least 1.

Reply:
Alright that makes sense.

I think we may consider changing this in the future. Otherwise if it happens that all nodes re-enroll at the same time, it will mean the random value is entirely predictable for that block height.

Reply:
The distance means the index of pre-images. So, the first pre-image(at distance 0) will be used at the height of the start of a cycle which is height 1. So there's no pre-image at distance 10 if the validator cycle is 10.

Please refer to this code.
```
 () @trusted {
    const ZeroDistance = 0;  // initial distance
    this.db.execute("INSERT INTO validator_set " ~
        "(key, cycle_length, enrolled_height, " ~
        "distance, preimage) VALUES (?, ?, ?, ?, ?)",
         enroll.utxo_key.toString(), enroll.cycle_length,
         block_height.value, ZeroDistance,
         enroll.random_seed.toString());
}();
```